enhancements/cluster-logging/forwarder-tagging.md

```diff
@@ -0,0 +1,132 @@
+---
+title: Forwarder adds Tags to Outbound Messages
+authors:
+  - "@alanconway"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-01
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Forwarder adds Tags to Outbound Messages
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Add configuration so the Forwarder can attach name+value fields to each message that is forwarded to a remote destination.
+For example: add a `datacenter-id` field with a value identifying a datacenter to every log message forwarded.
+
+Tags are associated with a forwarder `pipeline`. Different tags can be set based on `input` filters and `output` destinations. 
+For example `audit` logs can be tagged differently from `application` logs.
+
+This is a cluster-wide control point for _cluster administrators_, independent of the content of logs.
+These tags cannot be modified or deleted by applications in the local cluster.
+
+## Motivation
+
+### Goals
+
+* Cluster admin can tag outgoing messages at the forwarder.
+* Set fixed tags based on `output` destination.
+* Set fixed tags based on `input` origin, allow use of all present and future `input` filtering features.
+
+### Non-Goals
+
+* Tags are set at configuration, no computed tags based on message content.
+* Does not replace or interfere with other tagging mechanisms prior to the forwarder.
+
+## Proposal
+
+Add a `tags` map to the `pipeline` spec. The tags are applied to all logs
+passing through that pipeline.
+
+### User Stories [optional]
+
+#### As cluster admin I want to label all outgoing messages with a cluster ID
+
+```
+pipeline:
+  tags: { cluster-id: C1234 }
+  inputRefs: [application, infrastructure, audit]
+  outputRefs: [somewhere]
+```
+#### As cluster admin I want to label different message types differently.
+
+```
+pipeline:
+  tags: { logType: normal }
+  inputRefs: [application]
+  outputRefs: [somewhere]
+pipeline:
+  tags: { logType: special }
+  inputRefs: [infrastructure]
+  outputRefs: [somewhere]
+pipeline:
+  tags: { logType: secret }
+  inputRefs: [audit]
+  outputRefs: [somewhere]
+```
+
+### Risks and Mitigations
+
+Risk: Excessive use of tags may bloat outgoing log traffic.
+
+Mitigation: Cluster admins control use, with great power comes great responsibility.
+
+## Design Details
+
+Add a `tags` map to the `pipeline` spec.
+Forwarder includes tag directive in `fluentd` config.
+Syntax of new fluentd configuration: TBD
+
```

Comment:
+1 we need to define how we'll implement tags in this propossal. WIll give it some thought

Reply:
Should we ignore our own ES for now in this case? I mean, we introduce this as a mechanism for "outgoing" messages not for the ones we actually store in-cluster.

Reply:
I think not - we'll have to solve the problem for remote ES anyway, and we want our own ES to look the same as other ES. If we make it a special case it will be hard to swap it out in future.

Reply:
Right now, our code takes the map of labels and converts it into an array of string like `["foo=bar"]`  If we can optimally index as a map and not hose Elasticsearch then we should prefer that

Reply:
Have put a note in the proposal we should discuss.